rfcs/profiler.md

```diff
@@ -0,0 +1,144 @@
+
+![alt_text](images/flamegraph.png "image_tooltip")
+
+### Benefits
+
+- A major benefit of this design is that the profilers (both generic and custom ones) are not necessarily targeted toward single cell applications and can be used for any services across CZI.
+- Another benefit of this design is the fact that the generic profiler can be implemented and deployed in a reasonable amount of time opening the door for low-overhead perf/mem monitoring.
+- The third benefit of the design is that it can be easily extended by adding more custom profilers for different languages and software stack.
+
+### Drawbacks
+
+One drawback here is limiting custom profilers’ API to flamegraph. As mentioned earlier, there are many profilers with different output formats. One possible solution to this problem is to allow the byte array associated with each custom profiler in the DB schema to be open to different interpretations which while more scalable is a less secure solution.
+
+## Architecture
+
+![alt_text](images/profilerarchitecture.png "image_tooltip")
+
+As shown in the figure, the following steps take place as a part of the process of profiling:
+
+- First the generic profiler gets a request to run a process along with a group of custom process information (1).
+- As it starts the process using the **_time_** utility (2), it also starts the custom profilers by providing the process ID of the initiated process (3) so they can attach themselves to that process (4).
+- Once execution finishes,the **_time_** utility provides the time and performance metrics which are extracted with parsing (5) while the custom profilers provide **_flamegraphs_** of their metrics (5) and all that information plus some other fields get stored as a record in DB (5).
+
+## Metrics & Alerts
+
+A metric for performance and memory regression should be designed and implemented. The use of custom profilers can be conditional on detection of regression. The idea being once a regression is detected, we rerun the process with custom profilers to avoid slowing down normal execution.
+
+## Security & Data Privacy
+
+The central DB can pose a risk if that gets compromised as it reveals the history of all of the Single Cell Dataset runs.
+
+## Estimated Cost
+
+This design does not require adding new machines (nodes). It only relies on installation of the **_time_** utility and custom profilers across the nodes. The other cost is the storage needed for the central DB.
```

Comment:
Even if it doesn't belong to the `cost` section, we need to figure out the ownership of the centralized database. As far as I know, for the SOMA project, we opted to keep most of the assets in GitHub so that we don't tie them to e.g. a specific AWS account. This would be difficult if we decided to use a "real" database like DynamoDB.

Reply:
You are right. The final decision was to keep this on github. I'll update it